### ISSUE TITLE
CSVAnalysis does not update `upfiles` after a file is uploaded so the new file is not selectable

### ISSUE THREAD

**Issue by zygmuntszpak** [GenieFramework/Stipple.jl#166]
The current implementation of the CSVAnalysis demo is not quite right. When a user uploads a new file, the `upfiles` variable, which is part of `Stipple.select(:selected_file; options=:upfiles)` and is responsible for listing all available files, is not reactively updated.  

![upfiles not updated](https://user-images.githubusercontent.com/14922863/213333919-cde7b725-ff19-4b36-8281-dedcb2dad909.png)

The issue is that `onchangeany` is not set to trigger when upfiles changes, but only when a selected file or selected column changes. In practice, this means that a user has to first select a different column, which then triggers the handler, which then updates upfiles so that the new file appears (in this demo I made sure to first remove the `wineqeuality-red` and `wineqeuality-white` from the uploads directory so simulate a genuine upload of a new file). 

```julia
@handlers begin
    @onchangeany isready, selected_file, selected_column begin
        upfiles = readdir(FILE_PATH)
        data = CSV.read(joinpath(FILE_PATH, selected_file), DataFrame)
        columns = names(data)
        datatable = DataTable(data)
        if selected_column in names(data)
            irisplot = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
        end
    end
end
```

I have a simple fix for this, but I don't know how to do translate the fix into the new API.  The following implementation fixes the issue:

```julia
using Stipple
using StippleUI
using StipplePlotly
using CSV
using DataFrames 
using Random


const FILE_PATH = "uploads"
mkpath(FILE_PATH)

Genie.config.cors_headers["Access-Control-Allow-Origin"]  =  "*"
Genie.config.cors_headers["Access-Control-Allow-Headers"] = "Content-Type"
Genie.config.cors_headers["Access-Control-Allow-Methods"] = "GET,POST,PUT,DELETE,OPTIONS"
Genie.config.cors_allowed_origins = ["*"]

@vars AppModel begin
    title = "Limited Bandwidth Demo"
    selected_file = "iris.csv"
    selected_column = "petal.length"
    upfiles = readdir(FILE_PATH)
    columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
    irisplot = PlotData()
    files_dirty = false
end

function ui(model)
    page(
        model,
        class = "container",[
        row([
            cell(class="col-md-12", [
                                        uploader(label="Upload Dataset", accpt=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file")
                                    ])
        ]),
        row([
             cell(class="st-module", [
                                        h6("File")
                                        Stipple.select(:selected_file; options=:upfiles)
                                     ])
             cell(class="st-module", [
                                         h6("Column")
                                        Stipple.select(:selected_column; options=:columns)
                                     ])
        ]),
        row([
              cell(class="st-module", [
                                        h5("Histogram")
                                        plot(:irisplot)
                                     ])
            ])],
    )
end

function handlers(model)
    Stipple.onany(model.isready, model.selected_file, model.selected_column, model.upfiles) do isready, selected_file, selected_column, upfiles
        data = CSV.read(joinpath(FILE_PATH, selected_file), DataFrame)
        model.columns[] = names(data)
        if selected_column in names(data)
            model.irisplot[] = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
        end
    end
    model
end

model = Stipple.init(AppModel) |> handlers

route("/") do
  html(ui(model), context = @__MODULE__)
end

route("/", method = POST) do
    @show "Processing upload..."
    files = Genie.Requests.filespayload()
    for f in files
        write(joinpath(FILE_PATH, f[2].name), f[2].data)
    end
    if length(files) == 0
        @info "No file uploaded"
    end
    # Trigger upfiles change.
    model.upfiles[] =  readdir(FILE_PATH)

    return "Upload finished"
end

up()

```

Part of the reason it was relatively easy to fix the issue is that I had access to the explicitly defined `model`, and over time I have been able to piece together a vague mental model of the StippleUI workflow and the fact that the handlers update reactive variables. 

I think a drawback of the new UI is that, whilst it tries to abstract away some details and make things more readable, it further obscures the underlying mental model of the StippleUI workflow. If I hadn't already known that there were Observable variables powering the whole workflow and I first encountered this new API I think I would have struggled even more in fixing my own issues etc. 

For instance, the new API has the following route implementation for the CSVAnalysis

```julia
route("/", method = POST) do
  files = Genie.Requests.filespayload()
  for f in files
      write(joinpath(FILE_PATH, f[2].name), f[2].data)
  end
  if length(files) == 0
      @info "No file uploaded"
  end
  upfiles = readdir(FILE_PATH)
  return "Upload finished"
end
```

A newcomer may be perplexed why `upfiles` is not automatically updated. After all, the `upfiles` variable was decorated with `@out`. 
```julia
@out title = "CSV Analysis"
@out upfiles = readdir(FILE_PATH)
@in selected_file = "iris.csv"
@in selected_column = "petal.length"
@out columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
@out irisplot = PlotData()
```

I had to play around with `@macroexpand` to confirm that `@handlers` is injecting `Main_ReactiveModel.` in front of variables etc. The expansion shows things like `__model__` etc. which I can see are counterparts of `model = Stipple.init(AppModel) |> handlers` but because all this is hidden behind macros, I have no idea how to access this model variable, let alone ensure that the line 
```julia
upfiles = readdir(FILE_PATH)
```
transformed into something like `__model__.upfiles = readdir(FILE_PATH)`.

In summary I have both a question and some feedback. How would one go about introducing my "fix" using the new API? In terms of feedback, I just want to point out that the more work gets done by macros the harder it is to paint a picture of what is really going on, which in turn makes it harder for people to contribute to the codebase etc.

**Comment by PGimenez** [comment]
Hey @zygmuntszpak thanks for opening this issue. I recently created a similar one in https://github.com/GenieFramework/Stipple.jl/issues/163, so I moved the issue to the Stipple repository. We're currently looking into this.

As for your comment regarding `upfiles` not being updated in `route`, yes, that shouldn't have been there. I forgot to delete it after moving the update to the  `@onchangeany` block; fixing it now.

**Comment by hhaensel** [comment]
As promised, the main ingredients are:
```julia
function Base.notify(::AppModel, ::Val{:updatefiles}, args...)
    model.upfiles[] = readdir(FILE_PATH)
end

uploader(label="Upload Dataset", accpt=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file",
    var"@uploaded"="handle_event('', 'updatefiles')")
```
Together with a check whether the App looks like
```julia
using Stipple
using StippleUI
using StipplePlotly
using CSV
using DataFrames 
using Random


const FILE_PATH = "uploads"
mkpath(FILE_PATH)

Genie.config.cors_headers["Access-Control-Allow-Origin"]  =  "*"
Genie.config.cors_headers["Access-Control-Allow-Headers"] = "Content-Type"
Genie.config.cors_headers["Access-Control-Allow-Methods"] = "GET,POST,PUT,DELETE,OPTIONS"
Genie.config.cors_allowed_origins = ["*"]

@vars AppModel begin
    title = "Limited Bandwidth Demo"
    selected_file = "iris.csv"
    selected_column = "petal.length"
    upfiles = readdir(FILE_PATH)
    columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
    irisplot = PlotData()
    files_dirty = false
end

function ui(model)
    page(
        model,
        class = "container",[
        row([
            cell(class="col-md-12", [
                                        uploader(label="Upload Dataset", accpt=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file",
                                            var"@uploaded"="handle_event('', 'updatefiles')")
                                    ])
        ]),
        row([
             cell(class="st-module", [
                                        h6("File")
                                        Stipple.select(:selected_file; options=:upfiles)
                                     ])
             cell(class="st-module", [
                                         h6("Column")
                                        Stipple.select(:selected_column; options=:columns)
                                     ])
        ]),
        row([
              cell(class="st-module", [
                                        h5("Histogram")
                                        plot(:irisplot)
                                     ])
            ])],
    )
end

function handlers(model)
    Stipple.onany(model.isready, model.selected_file, model.selected_column, model.upfiles) do isready, selected_file, selected_column, upfiles
        filepath = joinpath(FILE_PATH, selected_file)
        isfile(filepath) || return
        data = CSV.read(filepath, DataFrame)
        model.columns[] = names(data)
        if selected_column in names(data)
            model.irisplot[] = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
        end
    end

    model
end

function Base.notify(::AppModel, ::Val{:updatefiles}, args...)
    model.upfiles[] = readdir(FILE_PATH)
end

model = Stipple.init(AppModel) |> handlers

route("/") do
  html(ui(model), context = @__MODULE__)
end

route("/", method = POST) do
    @show "Processing upload..."
    files = Genie.Requests.filespayload()
    for f in files
        write(joinpath(FILE_PATH, f[2].name), f[2].data)
    end
    if length(files) == 0
        @info "No file uploaded"
    end
    # Trigger upfiles change.
    model.upfiles[] =  readdir(FILE_PATH)

    return "Upload finished"
end

up()
```

**Comment by hhaensel** [comment]
Note that all users will see the same files. If you need to prevent that users see each other's files then you will have to either make a temp directory that is deleted upon garbage collection of the model or you have to identify the user by some mechanism and write all files to a user-specific directory.

**Comment by hhaensel** [comment]
If you translate this to the new ReactiveTools API, it would look like this:
```julia
using GenieFramework
using CSV
using DataFrames 
using Random

const FILE_PATH = "uploads"
mkpath(FILE_PATH)

Genie.config.cors_headers["Access-Control-Allow-Origin"]  =  "*"
Genie.config.cors_headers["Access-Control-Allow-Headers"] = "Content-Type"
Genie.config.cors_headers["Access-Control-Allow-Methods"] = "GET,POST,PUT,DELETE,OPTIONS"
Genie.config.cors_allowed_origins = ["*"]

@vars begin
    title = "Limited Bandwidth Demo"
    selected_file = "iris.csv"
    selected_column = "petal.length"
    upfiles = readdir(FILE_PATH)
    columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
    irisplot = PlotData()
    files_dirty = false
end

@onchange isready, selected_file, selected_column, upfiles begin
    filepath = joinpath(FILE_PATH, selected_file)
    isfile(filepath) || return
    data = CSV.read(filepath, DataFrame)
    columns = names(data)
    if selected_column in names(data)
        irisplot = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
    end
end

@event updatefiles begin
    upfiles = readdir(FILE_PATH)
end

@page("/", ui)

route("/", method = POST) do
    @show "Processing upload..."
    files = Genie.Requests.filespayload()
    for f in files
        write(joinpath(FILE_PATH, f[2].name), f[2].data)
    end
    if length(files) == 0
        @info "No file uploaded"
    end

    return "Upload finished"
end

function ui()
    [
        row([
            cell(class="col-md-12", [
                uploader(label="Upload Dataset", accept=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file",
                    var"@uploaded"="handle_event('', 'updatefiles')")
            ])
        ]),
        row([
            cell(class="st-module", [
                h6("File")
                Stipple.select(:selected_file; options=:upfiles)
            ])
            cell(class="st-module", [
                h6("Column")
                Stipple.select(:selected_column; options=:columns)
            ])
        ]),
        row([
            cell(class="st-module", [
                h5("Histogram")
                plot(:irisplot)
            ])
        ])
    ]
end

up()
```

**Comment by zygmuntszpak** [comment]
I'm a bit perplexed by this function:
```julia
function Base.notify(::AppModel, ::Val{:updatefiles}, args...)
    model.upfiles[] = readdir(FILE_PATH)
end
```

Since `model` is not explicitly passed into the function, how come it is accessible? 

I can see how `handlers` and `ui` can see model based on the route function:
```julia
route("/") do
  model = AppModel |> init |> handlers
  html(ui(model), context = @__MODULE__)
end
```

**Comment by hhaensel** [comment]
> I'm a bit perplexed by this function:
> 
> ```julia
> function Base.notify(::AppModel, ::Val{:updatefiles}, args...)
>     model.upfiles[] = readdir(FILE_PATH)
> end
> ```
> 
> Since `model` is not explicitly passed into the function, how come it is accessible?

That's wrong, it must read. 
```julia
function Base.notify(model::AppModel, ::Val{:updatefiles}, args...)
    model.upfiles[] = readdir(FILE_PATH)
end
```
I thought that I edited that part

**Comment by hhaensel** [comment]
But the new API doesn't need it

**Comment by hhaensel** [comment]
@zygmuntszpak 
Can we close this?

**Comment by zygmuntszpak** [comment]
I just had a go at trying to run the version which you translated into the new ReactiveTools API and it throws an error:
I ran the following example:
```julia
using GenieFramework
using CSV
using DataFrames 
using Random

const FILE_PATH = "uploads"
mkpath(FILE_PATH)

Genie.config.cors_headers["Access-Control-Allow-Origin"]  =  "*"
Genie.config.cors_headers["Access-Control-Allow-Headers"] = "Content-Type"
Genie.config.cors_headers["Access-Control-Allow-Methods"] = "GET,POST,PUT,DELETE,OPTIONS"
Genie.config.cors_allowed_origins = ["*"]

@vars begin
    title = "Limited Bandwidth Demo"
    selected_file = "iris.csv"
    selected_column = "petal.length"
    upfiles = readdir(FILE_PATH)
    columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
    irisplot = PlotData()
    files_dirty = false
end

@onchange isready, selected_file, selected_column, upfiles begin
    filepath = joinpath(FILE_PATH, selected_file)
    isfile(filepath) || return
    data = CSV.read(filepath, DataFrame)
    columns = names(data)
    if selected_column in names(data)
        irisplot = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
    end
end

@event updatefiles begin
    upfiles = readdir(FILE_PATH)
end

@page("/", ui)

route("/", method = POST) do
    @show "Processing upload..."
    files = Genie.Requests.filespayload()
    for f in files
        write(joinpath(FILE_PATH, f[2].name), f[2].data)
    end
    if length(files) == 0
        @info "No file uploaded"
    end

    return "Upload finished"
end

function ui()
    [
        row([
            cell(class="col-md-12", [
                uploader(label="Upload Dataset", accept=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file",
                    var"@uploaded"="handle_event('', 'updatefiles')")
            ])
        ]),
        row([
            cell(class="st-module", [
                h6("File")
                Stipple.select(:selected_file; options=:upfiles)
            ])
            cell(class="st-module", [
                h6("Column")
                Stipple.select(:selected_column; options=:columns)
            ])
        ]),
        row([
            cell(class="st-module", [
                h5("Histogram")
                plot(:irisplot)
            ])
        ])
    ]
end

up()
```

with the following package versions:

```julia
  [a59fdf5c] GenieFramework v1.8.0
[4acbeb90] Stipple v0.25.19
[a3c5d34a] StippleUI v0.21.2
```

and it throws the error: 

```julia
ERROR: LoadError: UndefVarError: @type not defined
in expression starting at C:\Users\work\.julia\packages\Stipple\0qknm\src\ReactiveTools.jl:795
in expression starting at C:\Users\work\.julia\packages\Stipple\0qknm\src\ReactiveTools.jl:795
```

Since the original tutorial by PGimenez was using the `@in @out` definitions, it might be worthwhile translating the `@vars` example into that notation as well. I tried to give it a go a moment ago, but then ran into this error.

**Comment by AbhimanyuAryan** [comment]
I got it let me post my code

**Comment by hhaensel** [comment]
What's the status here?

**Comment by AbhimanyuAryan** [comment]
Completely forgot. Will rebuild the demo tomorrow and post MWE. But I was able to get it working with what you suggested Helmut

**Comment by hhaensel** [comment]
Can we close this now?

**Comment by AbhimanyuAryan** [comment]
It works but I lost the demo code. But I used your code to build it to get the working demo. I think we can close this if @zygmuntszpak is good or no follow-up

**Comment by zygmuntszpak** [comment]
If you copy and paste the following code, does it work for you? 

```julia
using GenieFramework
using CSV
using DataFrames 
using Random

const FILE_PATH = "uploads"
mkpath(FILE_PATH)

Genie.config.cors_headers["Access-Control-Allow-Origin"]  =  "*"
Genie.config.cors_headers["Access-Control-Allow-Headers"] = "Content-Type"
Genie.config.cors_headers["Access-Control-Allow-Methods"] = "GET,POST,PUT,DELETE,OPTIONS"
Genie.config.cors_allowed_origins = ["*"]

@vars begin
    title = "Limited Bandwidth Demo"
    selected_file = "iris.csv"
    selected_column = "petal.length"
    upfiles = readdir(FILE_PATH)
    columns = ["petal.length", "petal.width", "sepal.length", "sepal.width", "variety"]
    irisplot = PlotData()
    files_dirty = false
end

@onchange isready, selected_file, selected_column, upfiles begin
    filepath = joinpath(FILE_PATH, selected_file)
    isfile(filepath) || return
    data = CSV.read(filepath, DataFrame)
    columns = names(data)
    if selected_column in names(data)
        irisplot = PlotData(x=data[!, selected_column], plot=StipplePlotly.Charts.PLOT_TYPE_HISTOGRAM)
    end
end

@event updatefiles begin
    upfiles = readdir(FILE_PATH)
end

@page("/", ui)

route("/", method = POST) do
    @show "Processing upload..."
    files = Genie.Requests.filespayload()
    for f in files
        write(joinpath(FILE_PATH, f[2].name), f[2].data)
    end
    if length(files) == 0
        @info "No file uploaded"
    end

    return "Upload finished"
end

function ui()
    [
        row([
            cell(class="col-md-12", [
                uploader(label="Upload Dataset", accept=".csv", multiple=true, method="POST", url="http://localhost:8000/", field__name="csv_file",
                    var"@uploaded"="handle_event('', 'updatefiles')")
            ])
        ]),
        row([
            cell(class="st-module", [
                h6("File")
                Stipple.select(:selected_file; options=:upfiles)
            ])
            cell(class="st-module", [
                h6("Column")
                Stipple.select(:selected_column; options=:columns)
            ])
        ]),
        row([
            cell(class="st-module", [
                h5("Histogram")
                plot(:irisplot)
            ])
        ])
    ]
end

up()
```

I am still encountering the following error:

```julia
ERROR: LoadError: UndefVarError: @type not defined
in expression starting at C:\Users\zygmu\.julia\packages\Stipple\Ird4L\src\ReactiveTools.jl:794
in expression starting at C:\Users\zygmu\.julia\packages\Stipple\Ird4L\src\ReactiveTools.jl:794
```

even with the updated package versions:
```julia
[a59fdf5c] GenieFramework v1.10.0
[4acbeb90]  Stipple v0.25.20
[a3c5d34a] StippleUI v0.21.2
```

Are you perhaps using some dev branch? I don't understand why I would be getting errors in `ReactiveTools` if all the package versions are synced up.

Note that I cannot update Stipple beyond the current version because of the following constraint:

```julia
Status `C:\Users\zygmu\.julia\dev\LimitedBandwidthDemo\Project.toml`
⌅ [4acbeb90] Stipple v0.25.20 (<v0.26.0): GenieDevTools, GenieFramework, StipplePlotly, StippleUI
```

So in my view, the issue is not quite resolved and I wouldn't close this yet.

**Comment by hhaensel** [comment]
You tested just during we were tagging new versions. StippleUI was not yet released. It does run for me with the latest versions.
```julia
(@v1.9) pkg> st Stipple StippleUI
Status `C:\Users\m136270\.julia\environments\v1.9\Project.toml`
  [4acbeb90] Stipple v0.26.0 `C:\Users\m136270\.julia\dev\Stipple`
  [a3c5d34a] StippleUI v0.22.0 `C:\Users\m136270\.julia\dev\StippleUI`
```

**Comment by zygmuntszpak** [comment]
I made another attempt at getting the above code fragment to work but still no luck. The first error I encountered was 
```julia
ERROR: UndefVarError: ui not defined
Stacktrace:
 [1] top-level scope
   @ C:\Users\zygmu\.julia\packages\Stipple\gGEUX\src\ReactiveTools.jl:749
```

I then moved the `ui` code function above the `@page("/", ui)` statement which made the code compile. 

However, the following event seems not to be firing: 

```julia
@event updatefiles begin
    @show "updating upfiles"
    upfiles = readdir(FILE_PATH)
end
```
when I upload a new CSV file. 

I'm on

```julia
 [4acbeb90] Stipple v0.26.1
 [a3c5d34a] StippleUI v0.22.0
 [a59fdf5c] GenieFramework v1.12.0
```

**Comment by hhaensel** [comment]
The reason seems to be that `@uploaded=var"@uploaded"="handle_event('', 'updatefiles')"` is not correctly parsed.
(@essenciary I am wondering why it is parsed at all, `ui()` is a `Vector{ParsedHTMLString}`)

There are two easy ways out:
- write `var"v-on:uploaded"="handle_event('', 'updatefiles')"` instead
- use the `@on` macro, which has a very clean syntax `@on(:uploaded, :updatefiles)`

... and putting `ui()` above the `@page` macro was the right move.

With these two changes the event is firing.

**Comment by zygmuntszpak** [comment]
Thank you very much. I used the `var"v-on:uploaded"="handle_event('', 'updatefiles')"` variant and can confirm that it is working as expected.
